gamecommands: add new `/remove` cheat-only command

nbusseneau opened the pull request:
This new command allows to remove units similar to `/destroy` except no death script is run (so no explosion nor wreckage).

It works in exactly the same way as `/destroy`:

- As unsynced, `/remove` will remove selected units.
- As synced, `/remove <unitID...>` will remove all unitIDs provided.

Due to the fact only the last call to `KillUnit` differs between `/destroy` and `/remove`, we refactor both to use the same base action executor and only differentiate them by description and a boolean flag.

nbusseneau commented:
![image](https://github.com/beyond-all-reason/spring/assets/4659919/b6f471d6-2d68-40db-a339-70878177a881)

Commands sent in order:
```
/destroy
/destroy 1
/destroy true
/destroy 0
/destroy false
```

Log:
```
[t=00:01:01.013528][f=0007067] [destroy] Killing units (leaveWrecks: true): 4235 13052 
[t=00:01:01.014642][f=0007067] Armada Commander is being attacked!
[t=00:01:11.059414][f=0013095] [destroy] Killing units (leaveWrecks: true): 18538 13434 26948 9661 
[t=00:01:17.090103][f=0016714] [destroy] Killing units (leaveWrecks: true): 14776 24691 23421 
[t=00:01:17.090530][f=0016714] Armada Commander is being attacked!
[t=00:01:24.174384][f=0020961] [destroy] Killing units (leaveWrecks: false): 16424 1775 11159 
[t=00:01:29.343248][f=0024064] [destroy] Killing units (leaveWrecks: false): 30600 10276 12662 
```

sprunk commented:
I somewhat dislike allowing 0/1 because these are valid unitIDs. "true" and "false" sound OK. How about a separate command instead though, one that conveys what happens? In `Spring.DestroyUnit` the "don't leave a wreck" parameter is called "reclaimed", so how about `/reclaim`? Or `/delete` or something like that.

WatchTheFort commented:
Does the existing `RemoveUnits` command fulfil the intended use cases for this?

sprunk commented:
That is not an existing engine command. A game-agnostic way to remove units without all the extra fuss is good.

6AKU66 commented:
> I somewhat dislike allowing 0/1 because these are valid unitIDs. "true" and "false" sound OK, but then what if somebody adds more bool parameters? How about a separate command instead, one that conveys what happens? In `Spring.DestroyUnit` the "don't leave a wreck" parameter is called "reclaimed", so how about `/reclaim`? Or `/delete`, `/remove` or something like that.

I vote for "remove", i think it's simplest to type one.

WatchTheFort commented:
So how are these commands related to `/luarules destroyunits` and `/luarules removeunits`?

sprunk commented:
`/luarules destroyunits` lets BAR admins run the command without enabling `/cheat`, otherwise it seems to be the same as `/destroy`.

`/luarules removeunits` would have the above relationship with the command being added in this PR.

nbusseneau commented:
I actually didn't know about `removeunits`. Due to both `destroyunits` and `removeunits` existing, I agree this is best and `/remove` should be the corresponding engine-side command. Will update the PR.

nbusseneau commented:
PTAL :)

sprunk commented on rts/Game/SyncedGameCommands.cpp:
```diff
@@ -195,33 +195,49 @@ class GiveActionExecutor : public ISyncedActionExecutor {
-			if (!argsStream)
-				break;
+		std::ostringstream oss;
+    std::copy(args.begin(), args.end(), std::ostream_iterator<std::string>(oss, " "));
+    std::string unitIDs(oss.str());
```
ws

nbusseneau commented:
Related: while doing this refactor I discovered that the last (and optional) parameter on `KillUnit()` is actually redundant, soooo #963.

nbusseneau commented on rts/Game/UnsyncedGameCommands.cpp:
```diff
@@ -3422,20 +3422,20 @@ class GiveActionExecutor : public IUnsyncedActionExecutor {
+		if (selectedUnitsHandler.selectedUnits.empty()) {
 			return false;
+		}
 
-		// kill selected units
 		std::stringstream ss;
-		ss << GetCommand();
-
-		for (const int unitID: selectedUnitsHandler.selectedUnits) {
+		ss << action.GetCmd();
+		for (const int unitID : selectedUnitsHandler.selectedUnits) {
```
The change from `this->GetCommand()` to `action.GetCmd()` is just to make it consistent.

For full details:
- `this->GetCommand()` returns the command as passed to executor, e.g. `Remove`.
- `action.GetCmd()` gets the actual command from the action itself, so e.g. `remove` if you typed `/remove` or `Remove` if you typed `/Remove`.

The executor is picked based on action command in case-insensitive way, so in the code I just made sure to use `action.GetCmd()` when passing the command from unsynced to synced (as it's supposed to be a passthrough), whereas I used `this->GetCommand()` for things related to executor logic (e.g. logging which executor we're in right now).

nbusseneau commented on rts/Game/SyncedGameCommands.cpp:
```diff
@@ -195,33 +195,49 @@ class GiveActionExecutor : public ISyncedActionExecutor {
 
-			if (!argsStream)
-				break;
+		std::ostringstream oss;
```
Correct. I forgot to remove this after changing the purpose the PR, since initially I had to skip the first argument... Will fix.

nbusseneau replied:
Fixed.

nbusseneau commented on rts/Game/SyncedGameCommands.cpp:
```diff
@@ -195,33 +195,49 @@ class GiveActionExecutor : public ISyncedActionExecutor {
+		LOG("[%s] unitIDs: %s", this->GetCommand().c_str(), unitIDs.c_str());
 
-			CUnit* unit = unitHandler.GetUnit(unitId);
+		for (auto it = args.begin(); it != args.end(); ++it) {
```
Good call -- same as above, in the initial PR we were skipping the first argument, so a foreach was not possible. Will fix.

nbusseneau replied:
Fixed.

nbusseneau commented on rts/Game/SyncedGameCommands.cpp:
```diff
@@ -195,33 +195,49 @@ class GiveActionExecutor : public ISyncedActionExecutor {
-		do {
-			argsStream >> unitId;
+	bool Execute(const SyncedAction& action) const {
+		const std::vector<std::string>& args = CSimpleParser::Tokenize(action.GetArgs(), 0);
```
Indeed, it seems we only have string-tokenization capabilities in Spring, at least as far as I could tell...

nbusseneau commented:
Changed the `minWords` parameter of `CSimpleParser::Tokenize` from `0` to `1` in the previous push by mistake (thereby breaking the commands when trying to remove/destroy more than 2 units at once), changed back to `0` as in the initial review (works fine again for any number of units).
